doc/explanation/package-management.md

```diff
@@ -0,0 +1,220 @@
+The history of the OCaml package ecosystem is quite long with a lot of projects
+building upon each other over many years. A significant step forward was the
+creation of an OCaml Package Manager, OPAM, along with the establishment of a
+public package repository which made it a lot more feasible to share code
+between people and projects.
+
+Over time best practices have evolved, and while OPAM has incorporated some
+changes, due to its existing user base and constraints it couldn't adopt all
+the modern workflows.
+
+Dune package management attempts to take the parts of the OPAM ecosystem that
+have stood the test of time and couple them with a modern workflow. Some of the
+improvements include:
+
+* Automatic package repository updates
+* Easily reproducible dependencies
+* All package dependencies declared in a single file that is kept in-sync
+* Per-project dependencies
+
+Dune plays well with the existing OCaml ecosystem and does not introduce a new
+type of packages - it uses the same package repository and Dune packages stay
+installable with OPAM.
+
+## Package Management in a Project
+
+This section describes what happens in a Dune project that is using the package
+management feature.
+
+## Dependency Selection
+
+The first step is to determine which packages need to be installed.
+Traditionally this has been defined in the `depends` field of a projects OPAM
+file(s).
+
+Since a while Dune has also supported {doc}`OPAM file generation
+</howto/opam-file-generation>` by specifying the package dependencies in the
+`dune-project`. Outside of this feautre, the `depends` stanza had not been used
+by Dune.
+
+The package management feature changes this, as Dune now determines the list of
+packages to install from the `depends` stanza in the `dune-project` file. This
+allows projects to completely omit generation of `.opam` files, as long as they
+use Dune for package management. Thus all dependencies are only declared in one
+file.
+
+For compatibility with a larger amount of existing projects Dune will also
+collect dependencies from `.opam` files in the project. So while recommended,
+there is no obligation to switch to declaring dependencies in the
+`dune-project`. Likewise the generation of `.opam` files will still work.
+
+## Locking
+
+To go from a set of dependency constraints for a project to a set of installed
+packages and versions there needs to be a step to determine the right packages
+and their versions to be installed.
+
+In `opam` this process happens as part of `opam install`, thus linking finding
+a solution that satifies the given constrains and installation into one step.
+Dune on the other hand moves finding a solution and installing into separate
+steps. First a solution is found and then packages are installed separately.
+The idea of finding a solution and recording it for later is popular in many
+package managers and is usually called locking.
```

Comment:
I think it can be nice to give some examples here

Reply:
I've added NPM, any other package managers worth mentioning? Cargo?

Reply:
I think NPM is a good one :+1: